plugins/authn/pwbased/pwbased.go

```diff
@@ -2,69 +2,74 @@ package pwbased
 	"aureole/internal/plugins/authn"
-	authnTypes "aureole/internal/plugins/authn/types"
 	authzTypes "aureole/internal/plugins/authz/types"
+	"aureole/internal/plugins/core"
 	"aureole/internal/plugins/pwhasher/types"
 	storageTypes "aureole/internal/plugins/storage/types"
+	"aureole/internal/router"
 	"fmt"
 	"github.com/mitchellh/mapstructure"
 	"path"
 )
 
 type pwBased struct {
-	rawConf        *configs.Authn
-	conf           *сonfig
-	projectContext *contextTypes.ProjectCtx
-	pwHasher       types.PwHasher
-	storage        storageTypes.Storage
-	identityColl   *collections.Collection
-	authorizer     authzTypes.Authorizer
+	rawConf      *configs.Authn
+	conf         *сonfig
+	pwHasher     types.PwHasher
+	storage      storageTypes.Storage
+	identityColl *collections.Collection
+	authorizer   authzTypes.Authorizer
 }
 
-func (p *pwBased) Initialize(appName string) error {
-	projectCtx := authn.Repository.ProjectCtx
-	adapterConf := &сonfig{}
-	if err := mapstructure.Decode(p.rawConf.Config, adapterConf); err != nil {
+func (p *pwBased) Initialize(appName string) (err error) {
+	p.conf, err = initConfig(&p.rawConf.Config)
+	if err != nil {
 		return err
 	}
-	adapterConf.setDefaults()
-
-	p.conf = adapterConf
-	p.projectContext = projectCtx
 
-	hasher, ok := projectCtx.Hashers[p.conf.MainHasher]
-	if !ok {
+	pluginsApi := authn.Repository.PluginsApi
+	p.pwHasher, err = pluginsApi.GetHasher(p.conf.MainHasher)
+	if err != nil {
 		return fmt.Errorf("hasher named '%s' is not declared", p.conf.MainHasher)
 	}
 
-	collection, ok := projectCtx.Collections[p.conf.Collection]
-	if !ok {
+	p.identityColl, err = pluginsApi.GetCollection(p.conf.Collection)
+	if err != nil {
 		return fmt.Errorf("collection named '%s' is not declared", p.conf.Collection)
 	}
 
-	storage, ok := projectCtx.Storages[p.conf.Storage]
-	if !ok {
+	p.storage, err = pluginsApi.GetStorage(p.conf.Storage)
+	if err != nil {
 		return fmt.Errorf("storage named '%s' is not declared", p.conf.Storage)
 	}
 
-	authorizer, ok := projectCtx.Apps[appName].Authorizers[p.rawConf.AuthzName]
-	if !ok {
+	p.authorizer, err = pluginsApi.GetAuthorizer(p.rawConf.AuthzName, appName)
+	if err != nil {
 		return fmt.Errorf("authorizer named '%s' is not declared", p.rawConf.AuthzName)
 	}
 
-	p.pwHasher = hasher
-	p.identityColl = collection
-	p.storage = storage
-	p.authorizer = authorizer
+	if err = p.storage.CheckFeaturesAvailable([]string{p.identityColl.Type}); err != nil {
+		return err
+	}
+
+	createRoutes(pluginsApi, p)
+	return err
+}
+
+func initConfig(rawConf *configs.RawConfig) (*сonfig, error) {
+	adapterConf := &сonfig{}
+	if err := mapstructure.Decode(rawConf, adapterConf); err != nil {
+		return nil, err
+	}
+	adapterConf.setDefaults()
 
-	return p.storage.CheckFeaturesAvailable([]string{p.identityColl.Type})
+	return adapterConf, nil
 }
 
-func (p *pwBased) GetRoutes() []authnTypes.Route {
-	return []authnTypes.Route{
+func createRoutes(pluginsApi *core.PluginsApi, p *pwBased) {
```

Comment:
Never pass pluginsApi as argument. We already globally have it in the Repository